Source/Core/DolphinQt2/NetPlay/NetPlayDialog.cpp

```diff
@@ -0,0 +1,549 @@
+#include <QLabel>
+#include <QLineEdit>
+#include <QListWidget>
+#include <QMessageBox>
+#include <QProgressDialog>
+#include <QPushButton>
+#include <QSpinBox>
+#include <QTextEdit>
+#include <QTimer>
+
+#include <sstream>
+
+#include "Common/CommonPaths.h"
+#include "Common/TraversalClient.h"
+#include "Core/ConfigManager.h"
+#include "Core/Core.h"
+#include "Core/NetPlayServer.h"
+#include "DolphinQt2/GameList/GameFile.h"
+#include "DolphinQt2/GameList/GameList.h"
+#include "DolphinQt2/NetPlay/GameListDialog.h"
+#include "DolphinQt2/NetPlay/NetPlay.h"
+#include "DolphinQt2/NetPlay/PadMappingDialog.h"
+#include "VideoCommon/VideoConfig.h"
+
+NetPlayDialog* NetPlayDialog::m_instance = nullptr;
+
+NetPlayDialog* NetPlayDialog::Instance()
+{
+  if (m_instance == nullptr)
+    m_instance = new NetPlayDialog;
+  return m_instance;
+}
+
+NetPlayDialog::NetPlayDialog()
+{
+  setWindowFlags(windowFlags() & ~Qt::WindowContextHelpButtonHint);
+
+  setWindowTitle(tr("Dolphin NetPlay"));
+
+  m_pad_mapping = new PadMappingDialog(this);
+
+  CreateChatLayout();
+  CreatePlayersLayout();
+  CreateMainLayout();
+  ConnectWidgets();
+
+  m_timer = new QTimer(this);
+  connect(m_timer, &QTimer::timeout, this, &NetPlayDialog::UpdateGUI);
+  m_timer->start(200);
+}
+
+void NetPlayDialog::CreateMainLayout()
+{
+  m_main_layout = new QGridLayout;
+  m_game_button = new QPushButton;
+  m_md5_box = new QComboBox;
+  m_start_button = new QPushButton(tr("Start"));
+  m_buffer_size_box = new QSpinBox;
+  m_save_sd_box = new QCheckBox(tr("Write save / SD-Card data"));
+  m_load_wii_box = new QCheckBox(tr("Load Wii Save"));
+  m_record_input_box = new QCheckBox(tr("Record inputs"));
+  m_button_box = new QDialogButtonBox(QDialogButtonBox::Close);
+  m_buffer_label = new QLabel(tr("Buffer:"));
+
+  for (const QString& text :
+       {tr("MD5 Check:"), tr("Current game"), tr("Other game"), tr("SD card")})
+    m_md5_box->addItem(text);
+
+  m_main_layout->addWidget(m_game_button, 0, 0, 1, 5);
+  m_main_layout->addWidget(m_md5_box, 0, 5);
+  m_main_layout->addWidget(m_chat_box, 1, 0, 1, 5);
+  m_main_layout->addWidget(m_players_box, 1, 5, 1, 1);
+  m_main_layout->addWidget(m_start_button, 2, 0);
+  m_main_layout->addWidget(m_buffer_label, 2, 1);
+  m_main_layout->addWidget(m_buffer_size_box, 2, 2);
+  m_main_layout->addWidget(m_save_sd_box, 2, 3);
+  m_main_layout->addWidget(m_load_wii_box, 2, 4);
+  m_main_layout->addWidget(m_record_input_box, 2, 5);
+  m_main_layout->addWidget(m_button_box, 3, 0, 1, -1);
+
+  setLayout(m_main_layout);
+}
+
+void NetPlayDialog::CreateChatLayout()
+{
+  m_chat_box = new QGroupBox(tr("Chat"));
+  m_chat_edit = new QTextEdit;
+  m_chat_type_edit = new QLineEdit;
+  m_chat_send_button = new QPushButton(tr("Send"));
+
+  m_chat_edit->setReadOnly(true);
+
+  auto* layout = new QGridLayout;
+
+  layout->addWidget(m_chat_edit, 0, 0, 1, -1);
+  layout->addWidget(m_chat_type_edit, 1, 0, 1, 4);
+  layout->addWidget(m_chat_send_button, 1, 4);
```

Comment:
Should be

```c++
layout->addWidget(m_chat_type_edit, 1, 0);
layout->addWidget(m_chat_send_button, 1, 1);
```

If one of those needs to be stretched out and Qt isn't doing it automatically, either set it's sizePolicy or the QGridLayout's columnStretch.